RFC inspec style guide

arlimus opened the pull request:
Introduce an InSpec style guide. 
Motivation found here: https://innovadom.io/articles/inspec-style

arlimus committed:
🚶 start style guide

arlimus committed:
more style guide on controls

arlimus committed:
different header style

arlimus committed:
added ruby style conventions

arlimus committed:
rewrite more usage of ruby docs

arlimus committed:
rename ruby_usage -> style in docs

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

jerryaldrichiii commented on docs/dsl_inspec.md:
```diff
@@ -252,3 +252,104 @@ control 'ssh-1' do
+# Using Ruby in InSpec
+
+The InSpec DSL is a Ruby based language. This allows you to be flexible with
+Ruby code in controls:
```
Do we normally add line breaks in our Markdown?

arlimus replied:
Good point, probably needs some cleanup. It renders all the same but i have to review the style on a few things ty!

jerryaldrichiii commented on docs/dsl_inspec.md:
```diff
@@ -252,3 +252,104 @@ control 'ssh-1' do
+  #require 'pry'; binding.pry;
+  describe perl_out do
+    its('exit_status') { should eq 0 }
+    its('stdout') { should match (/USE_64_BIT_ALL/) }
```
What are your thoughts on:

```
its('stdout') { should match /USE_64_BIT_ALL/ }
```

or

```
its('stdout') { should match(/USE_64_BIT_ALL/) }
```

arlimus replied:
Oh good point! we should probably get that included too 👍

jerryaldrichiii commented on docs/dsl_inspec.md:
````diff
@@ -252,3 +252,104 @@ control 'ssh-1' do
+methods available, etc. For the above example:
+
+```ruby
+[1] pry> perl_out.exit_status
````
I use `.result` a lot, up to you to use it in your example :smile:

jerryaldrichiii commented on docs/dsl_inspec.md:
```diff
@@ -252,3 +252,104 @@ control 'ssh-1' do
+=> ["/Library/Perl/5.18/darwin-thread-multi-2level",
+ "    /Library/Perl/5.18",
+...REDACTED...
+[5] pry> exit    # or abort
```
I prefer `exit!` over abort, but it is personal preference.

jerryaldrichiii commented on docs/style.md:
```diff
@@ -0,0 +1,188 @@
+# InSpec style guide
+
+This is the set of recommended InSpec styles you should use on your daily journey when writing controls.
```
Styles or style?

arlimus replied:
ouch, good catch!

jerryaldrichiii commented on docs/style.md:
```diff
@@ -0,0 +1,188 @@
+This is a subset of style conventions taken from [the Ruby style guide](https://github.com/rubocop-hq/ruby-style-guide):
+
+- Use: UTF-8 as source file encoding.
+- Use: 2 spaces per indentation level. No hard tabs
```
Could use a period on the end of this line.

jerryaldrichiii commented on docs/style.md:
```diff
@@ -0,0 +1,188 @@
+- Use: UTF-8 as source file encoding.
+- Use: 2 spaces per indentation level. No hard tabs
+- Use: Unix-style file endings.
+- Avoid: multiple expressions per line.
```
Could we have an example of this?

jerryaldrichiii commented on docs/style.md:
```diff
@@ -0,0 +1,188 @@
+- Use: 2 spaces per indentation level. No hard tabs
+- Use: Unix-style file endings.
+- Avoid: multiple expressions per line.
+- Use: Spaces around operators, after commas, colons and semicolons.
```
The lack of an oxford comma disturbes me :smile:

jerryaldrichiii commented on docs/style.md:
````diff
@@ -0,0 +1,188 @@
+when /centos/
+  include_controls 'centos-profile'
+...
+```
````
You have newlines in your other examples...not sure how it affects rendering though.

jerryaldrichiii commented on docs/style.md:
````diff
@@ -0,0 +1,188 @@
+- Set the `supports` attribute in `inspec.yml` files of the profile you want to include:
+```ruby
+supports:
+- platform-name: centos
````
Do we need indention here because YAML?

arlimus replied:
naw this should be fine :)

jerryaldrichiii commented on docs/style.md:
```diff
@@ -0,0 +1,188 @@
+Now whenever you run the base profile you can just `include_controls 'centos-profile'`.
+It will only run the included profiles is the platform matches the supported platform.
+
+
```
Extra newline :smile:

arlimus replied:
yeah i got those in a few place for readability of the MD file. Won't affect the rendered doc

jerryaldrichiii commented on docs/style.md:
````diff
@@ -0,0 +1,188 @@
+
+Use:
+```ruby
+control "test-all-files" do
````
Not sure how we stand on single vs double quotes...

arlimus replied:
That's a really good question and discussion point! Opinions?

arlimus replied:
(the ruby style guide is opinionated on this; the question is: do we put our users through this?)

jerryaldrichiii replied:
Answered in parent thread. TL;DR: I care too much :smile:

jerryaldrichiii commented on docs/style.md:
````diff
@@ -0,0 +1,188 @@
+
+```ruby
+describe command('whoami') do
+  its(:stdout) { should eq "bob\n" }
````
We use `its('stdout')` in all other examples.

arlimus replied:
good point!

jerryaldrichiii commented on docs/style.md:
```diff
@@ -0,0 +1,188 @@
+Reason: One of the best way to develop and explore tests is the interactive debugging shell `pry` (see the section on "Interactive Debugging with Pry" at the end of this page). However, after you finish your profile make sure you have no interactive statements included anymore. Sometimes interactive calls are hidden behind conditionals (`if` statements) that are harder to reach. These calls can e ...
+
+Avoid: `binding.pry` in production profiles
+Use: Use debugging calls during development only
```
Should we give an `Inspec::Log` example?

arlimus replied:
Good point!

jerryaldrichiii reviewed (commented):
I'm loving how this is starting...apologies for a whole bunch of comments instead of a proper review.

arlimus committed:
add style section for inner describe test blocks

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

jerryaldrichiii commented on docs/dsl_inspec.md:
```diff
@@ -285,18 +285,18 @@ control 'check-perl' do
     describe directory(path.strip) do
-      it { should_not be_writable.by('group') }
-      it { should_not be_writable.by('other') }
+      it { should_not be_writable.by 'group' }
```
Now see...I actually like `by('group')`...style is hard...

aaronlippold replied:
Well, my thought there is that if we have `.by('other')` and in other examples use the style `it { should match /joe/ }` and further `{ should be <= 3 }` vs `it { should_not be_writable.by 'group' }` then it would be hard to ping someone on `it { should match.(/joe/) }`. They would say why is the general style `x` in one place and `y` in another - I think anyway.

aaronlippold replied:
But I think what @jerryaldrichiii gut is reacting to is the `.` .... i.e. if we have made the interface `it { should_not be_writable_by 'group' }` then his normal habit of thinking of `.by('other')` as a function is is calling off of `be_writable` would serve him well. And really I can see his point in this gut instinct.

arlimus replied:
Yeah fully agree @aaronlippold , that was my problem too. I was pondering making an exception for chaining, but then it's tricky with the other argument. we do want to avoid parentheses on the simple cases + then be consistent, so it hit chaining in the process. But we could make an exception here...

jerryaldrichiii replied:
I think an exception should be made for chaining. I worry that people who are not developers will be confused by:

```
it { should be_writable.by 'group' }
```

because the `.` looks like a typo for a `_` to the uninitiated.

I much prefer making an exception for chaining and then adding aliases when possible (e.g. `be_writable_by`)

jerryaldrichiii commented on docs/style.md:
````diff
@@ -180,9 +199,15 @@ Developers may still use external gem dependencies but should vendor it
+Also you may find it helpful to use the inspec logging interface:
+
+```ruby
+Inspec::Log.info("Hi")
````
/me squints eyes at quote usage

I see where you stand :stuck_out_tongue_closed_eyes:

arlimus replied:
ahahahaha oh shoot.
i'm writing too much go these days

james-stocks commented on docs/style.md:
```diff
@@ -0,0 +1,213 @@
+
+### 5. Do not wrap controls in conditional statements
+
+Reason: This will create dynamic profiles whose controls depend on the execution. The problem here is that we cannot render the profile or provide its information before scanning a system. We want to be able to inform users of the contents of their profiles before they run them. It is valid to skip controls that are not necessary for a system, as long as you do it via `only_if` conditions. Ruby's ...
```
There's a similar issue if tests are wrapped inside an iterator

e.g. to test all files under /var/log

```
  command('find /var/log -type f').stdout.split("\n").each do |log_file|
    describe file(log_file) do
      it { should_not be_writable.by('group') }
      it { should_not be_executable.by('group') }
      it { should_not be_readable.by('other') }
      it { should_not be_writable.by('other') }
      it { should_not be_executable.by('other') }
    end
  end
```

This causes a "Control has no tests defined" warning from `inspec check`.

It would be awesome if there was some Inspec concept that replaced `each`. Or if `inspec check` could assume that the iterator is sensible and not attempt to evaluate it.

james-stocks replied:
I raised https://github.com/inspec/inspec/issues/3348 just this week regarding this - we currently can't fail on `inspec check` warnings in CI due to this warning being unavoidable.

arlimus committed:
fix double quotes

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

james-stocks commented on docs/style.md:
```diff
@@ -0,0 +1,213 @@
+Avoid: `controls/ssh_controls.rb`
+Use: `controls/ssh.rb`
+
+
```
Suggestion: 
Recommend that control authors do not put line breaks or indenting into the `desc` string.

Reason: Software that executes Inspec and presents the results needs to be able to render control code and execution results in a UI. The description for a control should not include line breaks or indenting for the description string that might conflict with formatting later in the UI.

e.g.

Avoid: 

```
desc <-- DESC
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
    Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
    Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
DESC
```

Use:
```
desc 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.'
```

Or use Ruby string separation to make a long string more readable in your text editor:

```
desc 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, '\
     'sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. '\
     'Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris '\
     'nisi ut aliquip ex ea commodo consequat. '\
     'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum '\
     'dolore eu fugiat nulla pariatur. '\
     'Excepteur sint occaecat cupidatat non proident, '\
     'sunt in culpa qui officia deserunt mollit anim id est laborum.'
```

TrevorBramble replied:
I'd prefer the `desc` method convert line break characters to spaces and condense down to single-spacing. The tools should conform to our needs, not the reverse, and HEREDOC is much more readable than the unbroken line or the concatenating strings.

james-stocks replied:
That sounds good!

arlimus committed:
rename to profile style guide

according to suggestion https://github.com/inspec/inspec/pull/3356\#issuecomment-422114114

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

arlimus committed:
remove ruby style portion

until we can conlude this part of the discussion https://github.com/inspec/inspec/pull/3356

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

clintoncwolfe commented on www/data/docs_sidebar.yml:
```diff
@@ -28,6 +28,8 @@ sidebar_links:
     link: "/docs/reference/reporters.html"
   - title: InSpec DSL
     link: "/docs/reference/dsl_inspec.html"
+  - title: Style guide
```
Consider labelling this the 'Profile Style Guide'.  We may wish to add a Resource Style Guide in the future, or a Ruby style guide.

arlimus replied:
Missed that part, thank you!

clintoncwolfe reviewed (approved):
I think this is a great start!  Thanks @arlimus!

arlimus committed:
rename everywhere to profile style guide

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

jquick reviewed (approved):
Looks good! Thanks @arlimus
